### PR TITLE
Feature/pass play improvements

### LINKED ISSUE #1011: Look into a daemon for computationally heavy algorithms
Look into using archipelago and islands from pagmo to make a daemon that can run heavy algorithms without impacting AI performance

### PR DESCRIPTION
**Pull request by RobotJente**
Fixes #1011 and the scoring algorithm for passing. Uses [pagmo](https://esa.github.io/pagmo2/) to calculate the best location for a pass receiver to stand.

- [x] Document code

- [x] Fix cmake

- [ ] Make receiving robot go to ball after it is passed

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -115,6 +116,40 @@ set(NEW_PLAYS_SOURCES
...
+        ${PROJECT_SOURCE_DIR}/src/conditions/ShouldHandleBall.cpp
+        ${PROJECT_SOURCE_DIR}/src/conditions/IsOnPassLine.cpp
+        ${PROJECT_SOURCE_DIR}/src/conditions/BallIsClose.cpp
+        ${PROJECT_SOURCE_DIR}/src/conditions/CanPlay.cpp
+        ${PROJECT_SOURCE_DIR}/src/conditions/RobotOutside.cpp
+        ${PROJECT_SOURCE_DIR}/src/conditions/RefStateIsNormalPlay.cpp
+        ${PROJECT_SOURCE_DIR}/src/conditions/RefBallIsMoving.cpp
+        ${PROJECT_SOURCE_DIR}/src/conditions/ResumePlayAfterPenalty.cpp
+        )
+
```
Comment (RobotJente) [inline CMakeLists.txt]: ?????? why are these here

File: include/roboteam_ai/stp/new_plays_analysis/PassProblem.h [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+        /// Pagmo fitness function, only used for pagmo, don't change the signature as Pagmo algo's expect this exact function
+        vector_double fitness(const vector_double &aDouble) const;
+
+        /// Actual fitness function
+        static double cost_function(const Vector2 &point, world_new::view::WorldDataView world, const world::Field& field);
+
+        // TODO: when this function is finished, refactor so it only uses those components of world it actually needs
+        void updateInfoForProblem(world_new::WorldData problemWorld, world::Field& field);
+
```
Comment (RobotJente) [inline include/roboteam_ai/stp/new_plays_analysis/PassProblem.h]:
```suggestion
```

File: src/stp/new_plays/Pass.cpp [hunk 3]
```diff
@@ -62,13 +60,19 @@ void Pass::calculateInfoForRoles() noexcept {
...
-    // TODO: is there really no better way to set data per role?
-    // Use this new information to assign the roles using the dealer.
-    // TODO: compute the passing position
-    const Vector2 passingPosition = Vector2(-2, -2);
+    const Vector2 passingPosition = Vector2{archipelago->get_champions_x()[0][0], archipelago->get_champions_x()[0][1]};
 
     // Calculate receiver info
-    if (stpInfos.find("pass_receiver") != stpInfos.end()) stpInfos["pass_receiver"].setPositionToMoveTo(passingPosition);
+    if (stpInfos.find("pass_receiver") != stpInfos.end()) {
+
```
Comment (RobotJente) [inline src/stp/new_plays/Pass.cpp]:
```suggestion
```

File: src/stp/new_plays/Pass.cpp [hunk 4]
```diff
@@ -106,6 +111,33 @@ std::vector<Vector2> Pass::calculateDefensivePositions(int numberOfDefenders, wo
...
+        return cand;
+    }
+
+    // Proximity of candidate and current locations:
+    if ((currentPosition - candidatePosition).length() > 0.1 *  field.getFieldLength()) {
+        return curr;
+    }
+
+    return cand;
+
```
Comment (RobotJente) [inline src/stp/new_plays/Pass.cpp]:
```suggestion
```

File: src/utilities/Dealer.cpp [hunk 5]
```diff
@@ -155,8 +155,10 @@ double Dealer::getDefaultFlagScores(const v::RobotView &robot, const Dealer::Dea
             LineSegment lineSegment = {world.getBall()->get()->getPos(), field->getOurGoalCenter()};
             return lineSegment.distanceToLine(robot->getPos());
         }
+
         case DealerFlagTitle::KEEPER:
             return costForProperty(robot->getId() == GameStateManager::getCurrentGameState().keeperId);
+
```
Comment (RobotJente) [inline src/utilities/Dealer.cpp]:
```suggestion
```

File: include/roboteam_ai/stp/new_plays_analysis/PassProblem.h [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+using namespace pagmo;
+namespace rtt::ai::stp {
+    class PassProblem : public pagmo::problem {
+        /**
+         * Fitness for a point on the field
+         * @param input
+         * @return
+         */
+    private:
+        //std::mutex world_mutex;
```
Comment (timovdk) [inline include/roboteam_ai/stp/new_plays_analysis/PassProblem.h]: Commented out code

File: include/roboteam_ai/stp/new_plays_analysis/PassProblem.h [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+#ifndef RTT_PASSPROBLEM_H
+#define RTT_PASSPROBLEM_H
+
+#include <pagmo/problem.hpp>
+#include "world_new/WorldData.hpp"
+#include "world_new/views/WorldDataView.hpp"
+#include "world/Field.h"
+using namespace pagmo;
+namespace rtt::ai::stp {
```
Comment (timovdk) [inline include/roboteam_ai/stp/new_plays_analysis/PassProblem.h]:
```suggestion
#include "world/Field.h"

using namespace pagmo;

namespace rtt::ai::stp {
```

File: include/roboteam_ai/stp/new_plays_analysis/PassProblem.h [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        // TODO: when this function is finished, refactor so it only uses those components of world it actually needs
+        void updateInfoForProblem(world_new::WorldData problemWorld, world::Field& field);
+
+        PassProblem() = default;
+
+    };
+}
+
+
+
```
Comment (timovdk) [inline include/roboteam_ai/stp/new_plays_analysis/PassProblem.h]:
```suggestion

```

File: src/ApplicationManager.cpp [hunk 9]
```diff
@@ -73,6 +82,11 @@ void ApplicationManager::start() {
 
             // publish settings, but limit this function call to only run 1 times/s at most
             t.limit([&]() { io::io.publishSettings(SETTINGS.toMessage()); }, 1);
+
+            auto end = std::chrono::steady_clock::now();
+            auto diff = std::chrono::duration<double>(end-begin);
+            std::cout << "tick duration: "<< diff.count() << std::endl;
+            std::cout << "tick allowed: " << 0.016 << std::endl;
```
Comment (timovdk) [inline src/ApplicationManager.cpp]: You should use the print utils for this i guess

File: src/stp/new_plays/Pass.cpp [hunk 10]
```diff
@@ -17,40 +13,42 @@ namespace rtt::ai::stp::play {
 Pass::Pass() : Play() {
     // TODO: decide start invariants
     startPlayInvariants.clear();
-    startPlayInvariants.emplace_back(std::make_unique<invariant::BallCloseToUsInvariant>());
+    //startPlayInvariants.emplace_back(std::make_unique<invariant::WeHaveBallInvariant>());
 
     // TODO: decide keep invariants
     keepPlayInvariants.clear();
-    keepPlayInvariants.emplace_back(std::make_unique<invariant::BallMovesSlowInvariant>());
+    //keepPlayInvariants.emplace_back(std::make_unique<invariant::BallMovesSlowInvariant>());
```
Comment (timovdk) [inline src/stp/new_plays/Pass.cpp]: You haven't set invariants yet

File: src/stp/new_plays/Pass.cpp [hunk 11]
```diff
@@ -17,40 +13,42 @@ namespace rtt::ai::stp::play {
...
-    flagMap.insert({"test_role_10", {closeToTheirGoalFlag}});
+    flagMap.insert({"defender1", {closeToTheirGoalFlag}});
+    flagMap.insert({"defender2", {closeToTheirGoalFlag}});
+    flagMap.insert({"defender3", {closeToTheirGoalFlag}});
+    flagMap.insert({"defender4", {closeToTheirGoalFlag}});
+    flagMap.insert({"defender5", {closeToTheirGoalFlag}});
+    flagMap.insert({"defender6", {closeToTheirGoalFlag}});
+    flagMap.insert({"defender7", {closeToTheirGoalFlag}});
+    flagMap.insert({"defender8", {closeToTheirGoalFlag}});
+    flagMap.insert({"defender9", {closeToTheirGoalFlag}});
```
Comment (timovdk) [inline src/stp/new_plays/Pass.cpp]: Maybe rethink these flags since having defenders close to their goal does not make sense in my humble opinion :)

File: src/stp/new_plays_analysis/PassProblem.cpp [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+// Created by jesse on 27-03-20.
+//
+
+#include <roboteam_utils/Print.h>
+#include <roboteam_utils/Vector2.h>
+#include <include/roboteam_ai/world_new/World.hpp>
+#include <utility>
+#include "stp/new_plays_analysis/PassProblem.h"
+using namespace pagmo;
+namespace rtt::ai::stp{
```
Comment (timovdk) [inline src/stp/new_plays_analysis/PassProblem.cpp]:
```suggestion
#include "stp/new_plays_analysis/PassProblem.h"

using namespace pagmo;

namespace rtt::ai::stp{
```

**Review by timovdk — CHANGES_REQUESTED** [review]
Some comments

File: include/roboteam_ai/stp/new_plays_analysis/PassProblem.h (hunk 6 above)
Comment (RobotJente) [inline include/roboteam_ai/stp/new_plays_analysis/PassProblem.h]:
```suggestion
```

File: include/roboteam_ai/stp/new_plays_analysis/PassProblem.h [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+        /**
+         * Fitness for a point on the field
+         * @param input
+         * @return
+         */
+    private:
+        //std::mutex world_mutex;
+
+        world_new::WorldData problemWorld{};
+
```
Comment (RobotJente) [inline include/roboteam_ai/stp/new_plays_analysis/PassProblem.h:25]:
```suggestion
```

File: include/roboteam_ai/stp/new_plays_analysis/PassProblem.h [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+         * Fitness for a point on the field
+         * @param input
+         * @return
+         */
+    private:
+
+        world_new::WorldData problemWorld{};
+        world::Field problemField;
+
+        static double shootSuccesReward(Vector2 point, world_new::view::WorldDataView world, const world::Field& field);
```
Comment (RobotJente) [inline include/roboteam_ai/stp/new_plays_analysis/PassProblem.h:30]:
```suggestion
        static double shootSuccesReward(Vector2 point, world_new::view::WorldDataView world, const world::Field& field);
        
```

File: include/roboteam_ai/stp/new_plays_analysis/PassProblem.h [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+        vector_double fitness(const vector_double &aDouble) const;
+
+        /// Actual fitness function
+        static double cost_function(const Vector2 &point, world_new::view::WorldDataView world, const world::Field& field);
+
+        // TODO: when this function is finished, refactor so it only uses those components of world it actually needs
+        void updateInfoForProblem(world_new::WorldData problemWorld, world::Field& field);
+
+        PassProblem() = default;
+
```
Comment (RobotJente) [inline include/roboteam_ai/stp/new_plays_analysis/PassProblem.h]:
```suggestion
```

File: include/roboteam_ai/stp/new_plays_analysis/PassProblem.h [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+using namespace pagmo;
+
+namespace rtt::ai::stp {
+    class PassProblem : public pagmo::problem {
+        /**
+         * Fitness for a point on the field
+         * @param input
+         * @return
+         */
```
Comment (timovdk) [inline include/roboteam_ai/stp/new_plays_analysis/PassProblem.h]: Where is the function?

File: src/stp/new_plays_analysis/PassProblem.cpp [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+#include <utility>
+#include "stp/new_plays_analysis/PassProblem.h"
+
+using namespace pagmo;
+
+namespace rtt::ai::stp{
+    vector_double PassProblem::fitness(const vector_double &dv) const {
+        auto score = 0.0;
+        auto point = Vector2(dv[0], dv[1]);
+        //RTT_DEBUG("HI IM HAPPENING")
```
Comment (timovdk) [inline src/stp/new_plays_analysis/PassProblem.cpp]: Commented out code

File: src/stp/new_plays_analysis/PassProblem.cpp [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+        auto xboundleft = problemField.getLeftmostX() + stp::control_constants::ROBOT_RADIUS;
+        auto yboundbottom = problemField.getBottommostY() + stp::control_constants::ROBOT_RADIUS;
+        auto yboundtop = problemField.getTopmostY() - stp::control_constants::ROBOT_RADIUS;
+
+        return {{xboundleft, yboundbottom}, {xboundright, yboundtop}};
+    }
+
+
+    void PassProblem::updateInfoForProblem(world_new::WorldData problemWorld, world::Field& field) {
+        //std::lock_guard<std::mutex> guard(world_mutex);
```
Comment (timovdk) [inline src/stp/new_plays_analysis/PassProblem.cpp]: Commented out code

File: include/roboteam_ai/stp/new_plays/Pass.h [hunk 19]
```diff
@@ -37,22 +37,27 @@ class Pass : public Play {
...
+     *      The proximity of the locations
+     *      The proximity of the robot to either location
+     * @param candidatePosition the position suggested by the current iteration of the pass algorithm
+     * @param currentPosition the position suggested by the previous iteration of the pass algorithm
+     * @return true if desirable to switch
+     */
+    std::pair<Vector2, double> compareNewLocationToCurrentLocation(Vector2 currentPosition, Vector2 candidatePosition);
+
+    Vector2 currentPassLocation;
+    double currentPassScore;
```
Comment (timovdk) [inline include/roboteam_ai/stp/new_plays/Pass.h:64]: Are you sure you initialize these before using them? if not you should default initialize. If yes, this is fine :)

File: include/roboteam_ai/ApplicationManager.h [hunk 20]
```diff
@@ -34,22 +40,34 @@ class ApplicationManager {
...
      * Function that decides whether to change plays given a world and field.
      * @param _world the current world state
      * @param field the current field state
      */
     void decidePlay(world_new::World* _world);
 
+    /**
+     * This function generates the plays vector
+     */
+    void setPlays();
```
Comment (timovdk) [inline include/roboteam_ai/ApplicationManager.h]: Looks like this function is not implemented

File: src/ApplicationManager.cpp [hunk 21]
```diff
@@ -172,12 +160,34 @@ void ApplicationManager::decidePlay(world_new::World *_world) {
...
 
     currentPlay->update();
     mainWindow->updatePlay(currentPlay);
 }
 
 ApplicationManager::ApplicationManager(ai::interface::MainWindow *mainWindow) { this->mainWindow = mainWindow; }
+
+
+
+void ApplicationManager::updateArchipelago() {
```
Comment (timovdk) [inline src/ApplicationManager.cpp]:
```suggestion
ApplicationManager::ApplicationManager(ai::interface::MainWindow *mainWindow) { this->mainWindow = mainWindow; }

void ApplicationManager::updateArchipelago() {
```

File: src/ApplicationManager.cpp [hunk 22]
```diff
@@ -172,12 +160,34 @@ void ApplicationManager::decidePlay(world_new::World *_world) {
...
+    /// generating pass problem.
+    auto pso = pagmo::pso(10, 0.6, 0.6, 0.6, 0.6, 4, 2, 2, 1, 0);
```
Comment (timovdk) [inline src/ApplicationManager.cpp]: What do these numbers mean?

File: src/ApplicationManager.cpp [hunk 23]
```diff
@@ -172,12 +160,34 @@ void ApplicationManager::decidePlay(world_new::World *_world) {
...
+
+    ai::stp::PassProblem passProblem{};
+
+    auto field = world_new::World::instance()->getField().value();
+    world_new::WorldData data = **world_new::World::instance()->getWorld();
+    passProblem.updateInfoForProblem(data, field);
+
+    auto passPopulation = pagmo::population(passProblem, 10, 0);
```
Comment (timovdk) [inline src/ApplicationManager.cpp]: 10 looks like a magic number

File: src/stp/new_plays/Pass.cpp [hunk 24]
```diff
@@ -42,68 +51,97 @@ Dealer::FlagMap Pass::decideRoleFlags() const noexcept {
...
+    auto cand = std::make_pair(candidatePosition, candidatePositionScore);
+    auto curr = std::make_pair(currentPosition, currentPositionScore);
+
+    // Always choose a score that is negative, if possible
+    if (currentPositionScore > 0 && (candidatePositionScore < 0)) {
+        return cand;
+    }
+
+    // Ratio of scores
+    if (candidatePositionScore/currentPositionScore > 1.7) {
```
Comment (timovdk) [inline src/stp/new_plays/Pass.cpp:135]: magic number

File: include/roboteam_ai/stp/new_plays/Pass.h (hunk 19 above)
Comment (RobotJente) [reply include/roboteam_ai/stp/new_plays/Pass.h:64]: Yes, in the constructor I do!

**Review by timovdk — APPROVED** [review]
